### PR TITLE
Fix wav2vec2 is_batched check to include 2-D numpy arrays

### LINKED ISSUE #22175: wav2vec processor batching logic is too restrictive
### System Info

transformers version at the time of writing is `4.26.1`

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
# !pip install transformers torch # in jupyter notebook
from transformers import Wav2Vec2Processor
import torch
import numpy as np

batch = 4

# create Wav2Vec2Processor
processor = Wav2Vec2Processor.from_pretrained("facebook/hubert-large-ls960-ft")
# generate random input tensor
input_tensor = torch.tensor(np.random.randn(batch, 10, 10))
# pass input tensor through processor
output = processor(input_tensor, return_tensors="pt")
print(output["input_values"].shape) # 1 x 4 x 10 x 10
```

### Expected behavior

It seems reasonable that an input could be of shape `batch x d_1 x d_2 ...` and I'd expect the output to have the same shape. However, [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py#L184) the code has an extra check for type list or tuple that results in it misinterpreting the input as a single example.

Side note: I'm unsure what to infer from the type checking logic because it doesn't match the type hints i.e. `tuple` isn't supposed to be possible here anyways, according to the `__call__` type hint. I did check some other examples of `is_batched` appearing in the `src/transformers/models` directory and they look similar but unexpected.

### PR DESCRIPTION
**Pull request by LWprogramming**
# What does this PR do?

Fixes #22175 so it treats 2-D numpy arrays as being batched.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sanchit-gandhi

### CODE REVIEW

**Comment by LWprogramming** [comment]
Note: I was having some trouble running the relevant test. For instance, `pytest tests/models/wav2vec2/test_feature_extraction_wav2vec2.py` fails with 

```
File ... path/to/file/transformers/src/transformers/training_args.py", line 67, in <module>
    from accelerate import PartialState
ImportError: cannot import name 'PartialState' from 'accelerate' (/Users/leonwu/opt/anaconda3/lib/python3.9/site-packages/accelerate/__init__.py)
```

I suspect this might be related to #22816, but wasn't sure if I should downgrade `transformers` itself if I'm trying to make a PR.

```
$ transformers-cli env

- `transformers` version: 4.29.0.dev0
- Platform: macOS-13.2.1-arm64-arm-64bit
- Python version: 3.10.10
- Huggingface_hub version: 0.14.1
- Safetensors version: not installed
- PyTorch version (GPU?): 2.0.1 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no
```

and my virtual environment's `accelerate` library is `0.19.0`.

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py [hunk 1]
```diff
@@ -181,9 +181,13 @@ def __call__(
                 "Failing to do so can result in silent errors that might be hard to debug."
             )
 
-        is_batched = bool(
-            isinstance(raw_speech, (list, tuple))
-            and (isinstance(raw_speech[0], np.ndarray) or isinstance(raw_speech[0], (tuple, list)))
+        is_batched_numpy = isinstance(raw_speech, np.ndarray) and len(raw_speech.shape) == 2
+        if is_batched_numpy:
+            assert (
+                len(raw_speech.shape) == 2
```
Comment (sanchit-gandhi) [inline src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:187]: Note that `is_batched_numpy` will **only** be True if `raw_speech` is 2-d based on the above line. So this assert will never be triggered! Probably what we can do is set `is_batched_numpy` based on whether `raw_speech` is a numpy array with dimensionality 2 or more:
```python
is_batched_numpy = isinstance(raw_speech, np.ndarray) and len(raw_speech.shape) > 1
```
And then check to make sure it has the correct dims:
```python
if is_batched_numpy:
    if len(raw_speech.shape) > 2:
        raise ValueError(f"Only mono-channel audio is supported for input to {self}"
```
(we usually go with ValueError's over asserts in transformers 🤗)

**Review by sanchit-gandhi — COMMENTED** [review]
Hey @LWprogramming! Thanks for jumping on this PR! The fix nearly looks ready - just one small suggestion to the check! Would it also be possible to add a test where we pass a batched numpy array and check that we get batched features out?

**Comment by sanchit-gandhi** [comment]
Regarding the venv issue you're facing, could you try to isolate it with:
```python
from transformers import training_args
```
I can't reproduce this error using `transformers` from main - could you try rebasing onto main to make sure it's not already been fixed?

**Comment by LWprogramming** [comment]
> Regarding the venv issue you're facing, could you try to isolate it with:
> 
> ```python
> from transformers import training_args
> ```
> 
> I can't reproduce this error using `transformers` from main - could you try rebasing onto main to make sure it's not already been fixed?

that one's fixed, but I ran into a different issue which looks quite a bit like https://github.com/huggingface/transformers/issues/18355#issuecomment-1200940810. I'm going to try the instructions there-- installation probably going to take a while :))

**Comment by LWprogramming** [comment]
Noting here that I needed to separately `pip install parameterized` for some reason, but I've added the tests and confirmed they work!

File: tests/models/wav2vec2/test_feature_extraction_wav2vec2.py [hunk 2]
```diff
@@ -123,6 +123,14 @@ def test_call(self):
         for enc_seq_1, enc_seq_2 in zip(encoded_sequences_1, encoded_sequences_2):
             self.assertTrue(np.allclose(enc_seq_1, enc_seq_2, atol=1e-3))
 
+        # Test 2-D numpy arrays are batched.
+        speech_inputs = [floats_list((1, x))[0] for x in (800, 800, 800)]
+        np_speech_inputs = [np.asarray(speech_input) for speech_input in speech_inputs]
```
Comment (sanchit-gandhi) [inline tests/models/wav2vec2/test_feature_extraction_wav2vec2.py:128]: This should all be a NumPy array to trigger our new condition `is_batched_numpy`! Otherwise we'll just trigger `is_batched` as before:
```suggestion
        np_speech_inputs = np.asarray(speech_inputs)
```
Note that we can call `np.asarray` on our entire `speech_inputs` list in one go since all the dims are the same

File: tests/models/wav2vec2/test_tokenization_wav2vec2.py [hunk 3]
```diff
@@ -164,6 +164,14 @@ def test_call(self):
         for enc_seq_1, enc_seq_2 in zip(encoded_sequences_1, encoded_sequences_2):
             self.assertTrue(np.allclose(enc_seq_1, enc_seq_2, atol=1e-3))
 
+        # Test 2-D numpy arrays are batched.
+        speech_inputs = [floats_list((1, x))[0] for x in (800, 800, 800)]
+        np_speech_inputs = [np.asarray(speech_input) for speech_input in speech_inputs]
```
Comment (sanchit-gandhi) [inline tests/models/wav2vec2/test_tokenization_wav2vec2.py:169]: Same here!

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py [hunk 4]
```diff
@@ -181,9 +181,12 @@ def __call__(
                 "Failing to do so can result in silent errors that might be hard to debug."
             )
 
-        is_batched = bool(
-            isinstance(raw_speech, (list, tuple))
-            and (isinstance(raw_speech[0], np.ndarray) or isinstance(raw_speech[0], (tuple, list)))
+        is_batched_numpy = isinstance(raw_speech, np.ndarray) and len(raw_speech.shape) > 1
+        if is_batched_numpy:
```
Comment (sanchit-gandhi) [inline src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:185]: This looks great 👌

**Review by sanchit-gandhi — APPROVED** [review]
Just one nit regarding the test, but otherwise LGTM! In a follow-up PR we can propagate the changes onto all of our audio feature extractors since this is a missing edge case for all of them, e.g. Whisper:
https://github.com/huggingface/transformers/blob/6a6225beab4ea51a6ac3e9cfd56e2b4cdedd0022/src/transformers/models/whisper/feature_extraction_whisper.py#L206-L209

**Comment by LWprogramming** [comment]
d'oh, I gotta be more careful with copilot generations! fixed

File: src/transformers/models/wav2vec2/tokenization_wav2vec2.py [hunk 5]
```diff
@@ -820,9 +820,12 @@ def __call__(
                 values, a list of numpy arrayr or a list of list of float values.
         """
 
-        is_batched = bool(
-            isinstance(raw_speech, (list, tuple))
-            and (isinstance(raw_speech[0], np.ndarray) or isinstance(raw_speech[0], (tuple, list)))
+        is_batched_numpy = isinstance(raw_speech, np.ndarray) and len(raw_speech.shape) > 1
+        if is_batched_numpy:
+            if len(raw_speech.shape) > 2:
+                raise ValueError(f"Only mono-channel audio is supported for input to {self}")
```
Comment (hollance) [inline src/transformers/models/wav2vec2/tokenization_wav2vec2.py:826]: Personally, I'd combine these two if-statements into one:

```python
if is_batched_numpy and len(raw_speech.shape) > 2:
    raise ValueError(...)
```

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py [hunk 6]
```diff
@@ -181,9 +181,12 @@ def __call__(
...
+            if len(raw_speech.shape) > 2:
+                raise ValueError(f"Only mono-channel audio is supported for input to {self}")
```
Comment (hollance) [inline src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:187]: I'm OK with throwing an error but something we might consider doing is mixing the audio to mono, which would be done by taking the mean over the channels. If I'm training a big model and one of my examples happens to be stereo, I'd rather that this automatically gets turned into mono (perhaps with a warning message) than fail.

**Review by hollance — COMMENTED** [review]
I think it's OK as-is, but did leave a couple of suggestions.

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py (hunk 6 above)
Comment (sanchit-gandhi) [reply src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:187]: Probably you'd want this mixing to happen outside of the feature extraction code an in your data loading code no? IMO the purpose of the feature extraction should be kept as streamlined as possible in mapping a mono input to input features / input values, extra data pre-processing / checks can happen in the data loading steps

Comment (hollance) [reply src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:187]: The way I see it, the feature extractor could easily take care of stuff like resampling and conversion to mono etc. But this is a design decision: if you say that it's up to the user to provide the feature extractor with data in exactly the correct input format, then that's OK with me too. It just makes the feature extractor less convenient and kind of assumes you'll always be using this in combination with something like `datasets` or have to write your own additional resampling / conversion code.

Comment (LWprogramming) [reply src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:187]: I think it's more logical to put in the data loader, but I wouldn't mind adding a bit to the docs here so it doesn't surprise anyone who's about to load a big dataset and crash things halfway through when a non-mono appears. How does sound?

**Comment by sanchit-gandhi** [comment]
Cool! This looks ready to me @LWprogramming 👍 Would you mind just running the quality fix up:
```
make style
```
And then pushing? This should fix the failing code quality test and re-trigger the CI

**Comment by LWprogramming** [comment]
Is there a way to try running tests non-locally besides Circle CI? The `examples_torch` is failing on a wav2vec thing but I'm unsure if the bf16 unexpected result is a problem with my code, and when I run it locally with `pytest --make-reports=examples_torch ./examples/pytorch/ | tee tests_output.txt` it looks extremely slow.

**Comment by sanchit-gandhi** [comment]
Failing test looks unrelated!

**Review by sgugger — APPROVED** [review]
Thanks for the fix!
